Fix duplicate people insertion

evan314159 opened the pull request:
Prevent duplicate records in Peoples by checking for existing records before bulk-inserting into Peoples and PeopleBaseItemMap.
    
Implements migration to cleanup existing duplicate records:
- Consolidates duplicate People records by (Name, PersonType)
- Updates PeopleBaseItemMap to point to canonical People records
- Removes duplicate mappings and orphaned records
- Makes PersonType column NOT NULL with unique constraint
- Creates unique index on (Name, PersonType)

JPVenson commented on Jellyfin.Server/Migrations/Routines/20250814000002_RemoveDuplicatePeopleStage2.cs:
The new Migration service allows mixing code and EFCore migrations.
For the naming you are correct.

JPVenson commented:
this also get fixed in #14795 by reintroducing a simple transaction over the People update operation. I honestly would prefer doing it that way as its more idomatic for handing this issue (and should also handle it way more performant with pgsql ;-) as that dbms can introduce a rangelock instead)

evan314159 commented:
OK, if your PR now doesn't generate duplicates then this one is not required and can be closed.

evan314159 commented:
#14795 still results in duplicate insertions into Peoples -- on a fresh install I had almost 260,000 duplicate entries in Peoples when it should only have around 9000 total records -- and does not cleanup previous duplicate insertions, so the PR is still relevant.

Rebased the change against #14795.  UpdatePeople in particular is identical with the Polly retry removed.  Added two new cases to the Stage 2 migration code:
- Step 2 added mappingsToCreate to create missing PeopleBaseItemMaps
- Step 5 remove orphaned People records (those with no BaseItem mappings)

These were picked up with the test cases that I had previously uploaded into one of the above comment -- attaching again [test_people_baseitems_consistency_sqlite.sql.txt](https://github.com/user-attachments/files/22448588/test_people_baseitems_consistency_sqlite.sql.txt).

Removing Polly does result in duplicate insert errors on Postgres however the data integrity on Postgres seems to be preserved i.e. the result passes the test cases.

Shadowghost commented:
> https://github.com/jellyfin/jellyfin/pull/14795 still results in duplicate insertions into Peoples -- on a fresh install I had almost 260,000 duplicate entries in Peoples when it should only have around 9000 total records -- and does not cleanup previous duplicate insertions, so the PR is still relevant.

Did you check on sqlite and/or pgsql?

crobibero commented:
People de-duplication is handled in https://github.com/jellyfin/jellyfin/pull/14848

This doesn't strictly prevent it but cleans up during validation.
